File: divide-two-integers/divideTwoIntegers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        if divisor == 0:
            raise ValueError("Invalid input")
        MAX_INT = (1 << 31) - 1
        MIN_INT = -(1 << 31)
        
        res = 0
        positive = (dividend < 0) is (divisor < 0)
        dividend = abs(dividend)
        divisor = abs(divisor)
        while dividend >= divisor:
            tmp, i = divisor, 1
            while dividend >= tmp:
                dividend -= tmp
                res += i
                if (positive and res > MAX_INT) or (not positive and res < MIN_INT):
                    return MAX_INT
                tmp <<= 1
                i <<= 1
        
        if positive:
            return res
        else:
            return -res

File: divide-two-integers/divideTwoIntegers.java
public class Solution {
    public int divide(int dividend, int divisor) {
        if(divisor == 0) {
            throw new IllegalArgumentException("Invalid input");
        }
        long res = 0;
        int sign = ((dividend < 0) == (divisor < 0)) ? 1 : -1;
        long a = Math.abs((long)dividend), b = Math.abs((long)divisor);
        while(a >= b) {
            long tmp = b, i = 1;
            while(a >= tmp) {
                a -= tmp;
                res += i;
                if((sign == 1 && res > Integer.MAX_VALUE) || (sign == -1 && -res < Integer.MIN_VALUE)) {
                    return Integer.MAX_VALUE;
                }
                tmp <<= 1;
                i <<= 1;    
            }
        }
        return (int)res * sign;
    }
}